ntoskrnl/config/ntapi.c

```diff
@@ -1452,23 +1452,516 @@ NtLockRegistryKey(IN HANDLE KeyHandle)
+    KeEnterCriticalRegion();
+
+    /* Block registry notifications */
+    CmpLockNotifyExclusive();
+
+    /* Check for user-mode caller */
+    if (PreviousMode != KernelMode)
+    {
+        _SEH2_TRY
+        {
+            ProbeForWrite(IoStatusBlock, sizeof(IO_STATUS_BLOCK), sizeof(ULONG));
+            ProbeForWrite(Buffer, Length, sizeof(ULONG));
+
+            /* Probe and capture subordinate objects */
+            if (Count > 0)
+            {
+                LocalSubObjects = ExAllocatePoolZero(NonPagedPool, Count * sizeof(OBJECT_ATTRIBUTES), TAG_CM);
+                if (!LocalSubObjects)
+                {
+                    ObDereferenceObject(KeyObject);
+                    Status = STATUS_INSUFFICIENT_RESOURCES;
+                    _SEH2_YIELD(goto Quit);
+                }
+
+                SubNames = ExAllocatePoolZero(NonPagedPool, Count * sizeof(UNICODE_STRING), TAG_CM);
+                if (!SubNames)
+                {
+                    ObDereferenceObject(KeyObject);
+                    ExFreePoolWithTag(LocalSubObjects, TAG_CM);
+                    Status = STATUS_INSUFFICIENT_RESOURCES;
+                    _SEH2_YIELD(goto Quit);
+                }
+
+                for (int i = 0; i < Count; i++)
+                {
+                    Status = ProbeAndCaptureObjectAttributes(&LocalSubObjects[i],
+                                                             &SubNames[i],
+                                                             PreviousMode,
+                                                             &SubordinateObjects[i],
+                                                             FALSE);
+                    if (!NT_SUCCESS(Status))
+                        break;
+                }
+            }
+        }
+        _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+        {
+            /* Return the exception code */
+            Status = _SEH2_GetExceptionCode();
+        }
+        _SEH2_END;
+
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("NtNotifyChangeMultipleKeys: Probing buffers failed. (0x%lx)\n", Status);
+            if (KeyObject)
+                ObDereferenceObject(KeyObject);
+            if (LocalSubObjects)
+                ExFreePoolWithTag(LocalSubObjects, TAG_CM);
+            if (SubNames)
+                ExFreePoolWithTag(SubNames, TAG_CM);
+            goto Quit;
+        }
+    }
+
+    /* Open subordinate objects */
+    if (Count > 0)
+    {
+        LocalSubObjectsKeyBody = ExAllocatePoolZero(NonPagedPool, Count * sizeof(PCM_KEY_BODY), TAG_CM);
+        if (!LocalSubObjectsKeyBody)
+        {
+            Status = STATUS_INSUFFICIENT_RESOURCES;
+            goto Failure;
+        }
+
+        for (int i = 0; i < Count; i++)
+        {
+            /* Open a kernel-mode handle, since we're using it internally without returning it to the user */
+            HANDLE SubHandle;
+            LocalSubObjects[i].Attributes |= OBJ_KERNEL_HANDLE;
```

Comment:
This is never allocated, if the call came from kernel mode. What you can do is:
```
    if (PreviousMode != KernelMode)
    {
        // Allocate / probe / capture
        SubordinateObjects = LocalSubObjects;
    }
```
And subsequently only use `SubordinateObjects`. Similar for the `_Out_` parameters.